[gitsac] iP

gitsac opened the pull request:
# **Knight**
"Now, your mind can forget what's to be done, for I will remember it for you." - Knight

Knight frees you from life's smallest worries, by remembering all the tasks you need to do. It's:

- text-based
- easy to learn
- incredibly efficient when proficient

## **Setting you up for success**

Ready to remember less things, and start doing more?
All you need to do is,

- download it from [here](https://github.com/gitsac/ip/releases/tag/A-Jar).
- open a command window in the same directory
- using the command window, type the following command:
`java -jar Duke.jar`
- add your tasks.
- let it manage your tasks for you 😉

And it is FREE!

## **Features:**

- [x]  Managing tasks
- [x]  Managing deadlines
- [ ]  Reminders (coming soon)

If you happen to do a little coding in Java, see if this makes sense to you:
```
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

gitsac committed:
Add level-1 Increment

gitsac committed:
Removed class file

gitsac committed:
Add level-2 Increment

gitsac committed:
Add level-3 Increment

gitsac committed:
Add level-4 Increment

gitsac committed:
Abstracted tasks into their own classes, and fixed formatting for task output when using list command

gitsac committed:
Added level-5 Increment

gitsac committed:
Added level-6 Increment

gitsac committed:
Added automated text ui testing and re-formatted
text output of certain messages

gitsac committed:
Edited formatting of message output for tasks

gitsac committed:
Add level-7 Increment

gitsac committed:
Edit saved data file directory

gitsac committed:
Add Level-8 Increment

gitsac committed:
Merge branch 'branch-Level-7'

gitsac committed:
merge conflict resolved

gitsac committed:
Merge branch 'branch-Level-8' with merge conflict resolved

gitsac committed:
Merge branch 'master' of https://github.com/gitsac/ip

gitsac committed:
Add A-MoreOOP Increment

gitsac committed:
Add A-Packages Increment

gitsac committed:
Merge remote-tracking branch 'origin/add-gradle-support'

gitsac committed:
Add A-Gradle Increment

gitsac committed:
Remove commented-out lines

gitsac committed:
Remove more commented-out lines no longer in use

gitsac committed:
Fix function obtaining date of deadline task

gitsac committed:
Add A-JUnit Increment

gitsac committed:
Add A-Jar Increment

gitsac committed:
Add Level-9 Increment

gitsac committed:
Remove redundant lines of code and added additional abstraction when getting class fields

gitsac committed:
Add JavaDocs for all relevant methods

gitsac committed:
Fix function obtaining date of deadline task

gitsac committed:
Merge branch 'branch-A-JavaDoc'

# Conflicts:
#	src/main/java/Duke.java
#	src/main/java/duke/Parser.java
#	src/main/java/duke/TaskList.java
#	src/main/java/duke/Ui.java
#	src/main/java/duke/tasktypes/Deadlines.java

wz2k commented on src/main/java/data/dukedata.txt:
```diff
@@ -0,0 +1,3 @@
+1.[T][X] test2              
+2.[T][X] foru    
+3.[D][X] forum enjoyer (by: Tuesday) 
```
You seem to have 2 data text file and folder. Just a reminder to double check you file path and delete the redundant folder

wz2k commented on src/main/java/duke/DukeExceptions.java:
```diff
@@ -0,0 +1,38 @@
+    protected String printThisOut;
+
+    /**
+     * Constructor to initialize an instance of DukeException
+     * @param commandType String which indicates type of DukeException
```
There should be a empty line between description and parameter/return/throws section. This is relevant to all your javadoc comments.

wz2k commented on src/main/java/duke/DukeExceptions.java:
```diff
@@ -0,0 +1,38 @@
+    @Override
+    public String toString() {
+        String toReturn = ":( Sorry, this is not a supported command!";
+        if (printThisOut.equals("Wrong size for mark/unmark")) {
+            toReturn = ":( Sorry, the number input is wrong. Please check the possible indexes again using list!";
+        }
+        if (printThisOut.equals("deadline") || printThisOut.equals("todo")) {
+            toReturn = ":( Sorry, the description of a " + this.printThisOut + " cannot be empty!";
+        }
+        if (printThisOut.equals("event")) {
+            toReturn = ":( Sorry, the description of an " + this.printThisOut + " cannot be empty!";
+        }
+        if (printThisOut.equals("find")) {
+            toReturn = ":( Sorry, please input a keyword!";
+        }
```
Possible to use a switch statement here with fall through for deadline and todo but this is up to you.

wz2k commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,207 @@
+
+            if (input.startsWith("find")) {
+                handleFindTask(input, listOfTasks);
+                return;
+            }
+
+            if (input.startsWith("deadline")) {
+                handleDeadlineTask(input, listOfTasks);
+                return;
+            }
+
+            if (input.startsWith("event")) {
+                handleEventTask(input, listOfTasks);
+                return;
+            }
+
+            if (input.equals("list")) {
+                handleList(listOfTasks);
+                return;
+            }
+
+            if (input.startsWith("delete")) {
+                handleDelete(input, listOfTasks);
+                return;
+            }
+
+            if (input.startsWith("checkdue")) {
+                handleCheckDue(input, listOfTasks);
+                return;
+            }
+
+            if (input.startsWith("mark")) {
+                handleMark(input, listOfTasks);
+                return;
+            }
+
+            if (input.startsWith("unmark")) {
+                handleUnmark(input, listOfTasks);
+                return;
+            }
```
One way to use 'startsWith' together with a switch statement is to use the split method.

wz2k commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,207 @@
+     */
+    public void handleDeadlineTask(String input, TaskList listOfTasks) throws DukeExceptions {
+        try {
+            String useForInit = input.substring(8);
+            Task toAdd = new Deadlines(useForInit);
```
Perhaps instead of useForInit, the variable name can describe the name eg. deadlineDescription

wz2k commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,182 @@
+     * @throws DukeExceptions
+     */
+    public void handleDeadline(Character isDone, ArrayList<Task> listOfTasks, String requiredInformation) throws DukeExceptions {
+                String deadlineWithBy = requiredInformation.substring(requiredInformation.lastIndexOf("("));
+                String rawDate = deadlineWithBy.split("\\(by: ")[1].split("\\)")[0];
+                String[] nameSplitInArr = requiredInformation.substring(6).split(" ");
+                String nameOfTask = nameSplitInArr[0];
+                for (int i = 1; i < nameSplitInArr.length; i++) {
+                    if (nameSplitInArr[i].equals("(by:")) {
+                        break;
+                    }
+                    nameOfTask = nameOfTask + " " + nameSplitInArr[i];
+                }
+                String toInitialize = nameOfTask + " /by" + " " + rawDate;
+                Task toAdd = new Deadlines(toInitialize);
+                if (isDone.equals('X')) {
+                    toAdd.setDone();
+                }
+                listOfTasks.add(toAdd);
```
Take note of indentation error.

wz2k commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,182 @@
+     * @throws DukeExceptions
+     */
+    public void handleEvents(Character isDone, ArrayList<Task> listOfTasks, String requiredInformation) throws DukeExceptions {
+                String timeframe = requiredInformation.substring(requiredInformation.lastIndexOf("("));
+                String from = timeframe.substring(1).split(" to:")[0].substring(5);
+                String to = timeframe.substring(1).split("to: ")[1].split("\\)")[0];
+                String[] nameSplitInArr = requiredInformation.substring(6).split(" ");
+                String nameOfTask = nameSplitInArr[0];
+                for (int i = 1; i < nameSplitInArr.length; i++) {
+                    if (nameSplitInArr[i].equals("(from:")) {
+                        break;
+                    }
+                    nameOfTask = nameOfTask + " " + nameSplitInArr[i];
+                }
+                String toInitialize = nameOfTask + " /from" + from + " /to " + to;
+                Task toAdd = new Events(toInitialize);
+                if (isDone.equals('X')) {
+                    toAdd.setDone();
+                }
+                listOfTasks.add(toAdd); 
```
Take note of indentation error.

wz2k commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,142 @@
+                throw new DukeExceptions("Wrong size for mark/unmark");
+            }
+            Task gettingTask = listOfTasks.remove(indexToUse);
+            String toOutput = "Noted. I've removed this task:\n  " + gettingTask.toString() + "\nNow you have " + listOfTasks.size() + " tasks in the list";
```
This line too long. Possible to break. To not spam comments, I would not comment on other long lines but try to keep line lengths to 110 characters.

wz2k commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,31 @@
+package duke;
+
+import java.io.BufferedReader;
+import java.io.InputStreamReader;
+import java.io.IOException;
+
+/**
+ * Class used to read user's input for Duke chatbot.
+ */
+public class Ui {
+    
+    protected BufferedReader readingInput;
+
+    /**
+     * Constructor to initiate a bufferedreader to get user input.
+     */
+    public Ui() {
+        this.readingInput = new BufferedReader(new InputStreamReader(System.in));
+    }
+
+    /**
+     * Function to read user input.
+     * @return String representation of user input.
+     * @throws IOException
+     */
+    public String gettingUserInput() throws IOException {
+        String userInput = readingInput.readLine();
+        return userInput;
+    }
+
+}
```
Perhaps all you printing of messages could also be handled by the UI class

wz2k commented on src/main/java/duke/tasktypes/Deadlines.java:
```diff
@@ -0,0 +1,200 @@
+/**
+ * Class which represents a Task with a deadline.
+ */
+public class Deadlines extends Task {
```
Perhaps it is better to name is in singular form as each object from this class is a singular deadline.

wz2k commented on src/main/java/duke/tasktypes/Deadlines.java:
```diff
@@ -0,0 +1,200 @@
+    public String toString() {
+        String toReturn = "";
+        DateTimeFormatter newFormat = DateTimeFormatter.ofPattern("MMM/d/uuuu");
+        if (this.dueDateBy == null) {
+            String[] checkIfTrulyNull = this.endsBy.split(" ");
+            if (checkIfTrulyNull.length >= 3) {
+                String checkingInString = checkIfTrulyNull[0] + "/" + checkIfTrulyNull[1] + "/" + checkIfTrulyNull[2];
+                if (checkValidityOfDateFromList(checkingInString)) {
+                    reformat();
+                    if (checkIfTrulyNull.length == 4) {
+                        formatTimeIfValid(checkIfTrulyNull[3]);
+                    }
+                } else {
+                    System.out.println("The date is invalid!");
+                }
+            }
+        }
```
The module wants us to avoid deep nesting so you might want to change this if time permits

wz2k commented on src/main/java/duke/tasktypes/Deadlines.java:
```diff
@@ -0,0 +1,200 @@
+     * @return String representation of deadline task.
+     */
+    @Override
+    public String toString() {
```
Try to keep methods to a limit of 30 lines if possible

wz2k commented on src/main/java/duke/tasktypes/Task.java:
```diff
@@ -0,0 +1,31 @@
+ */
+public class Task {
+    protected String name;
+    protected boolean done;
```
Perhaps isDone would be better as a boolean variable name

wz2k reviewed (commented):
Overall, I can see that you have put in a lot of efforts to verify inputs. Just need to address some small changes to your code. Keep up the good work! :thumbsup:

Sheemo commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,71 @@
+            System.out.println(DE.toString());
+        }
+    }
+
```
Exception variable name should still follow coding style

Sheemo commented on src/main/java/duke/DukeExceptions.java:
```diff
@@ -0,0 +1,38 @@
+ */
+public class DukeExceptions extends Exception {
+    protected String printThisOut;
+
```
Better to use a variable name like `toPrint`

Sheemo replied:
Also better to call your exception `DukeException`

Sheemo commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,207 @@
+import duke.tasktypes.Events;
+import duke.tasktypes.ToDo;
+import duke.tasktypes.Task;
+
```
Great use of packages!

Sheemo commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,182 @@
+        scannerForFileData.close();
+        return useThis;
+    }
+
```
Might need to consider corrupted files and handle them gracefully

Sheemo commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,142 @@
+            toPrint = "Got it. I've added this task:\n  " + toAdd.toString() + "\nNow you have " + listOfTasks.size() + " tasks in the list.";                   
+        }
+        System.out.println(toPrint);
+    }
```
Printing listOfTasks.size() should be abstracted out to a function. Now there is a lot of repeated code. The function could also take in a boolean parameter isPlural for that extra "s" case

Sheemo commented on src/main/java/duke/tasktypes/Task.java:
```diff
@@ -0,0 +1,31 @@
+        return this.name;
+    }
+
+}
```
Task should be responsible for printing the [X] <description> as this is repeated across all subclasses

gitsac committed:
Fix code formatting according to checkstyle

gitsac committed:
Add JavaFX tutorial parts 1-3

gitsac committed:
Add basic JavaFX functionality

gitsac committed:
Refactor classes into appropriate packages

gitsac committed:
Add JavaDocs for GUI classes and fixed checkstyle

gitsac committed:
Edit variable to successfully build with GUI

gitsac committed:
Add assertions to Duke

In many methods, it is assumed that certain variables or objects
are of a certain state.

By using assertions, code quality can be improved to make those
assumptions a guarantee.

Assumptions make it difficult to debug if the error
stack trace is very long.

It is important to hold assumptions in current state as otherwise
there will be a need to check for many things at all methods.

Let's improve code quality and make debugging easier by adding
assumptions.

gitsac committed:
Fix JavaDocs formatting

gitsac committed:
Improve code quality by refactoring

The code body is functioning well but makes for poor reading at
times due to issues like big chunks of code and deep nesting.

By abstracting deep nesting and big chunks of code
into multiple helper functions, the code becomes more readable.

Let's practice SLAP and be mindful of readability in the future.

gitsac committed:
Merge pull request #2 from gitsac/branch-A-Assertions

Add assertions to Duke

gitsac committed:
Merge branch 'master' into branch-A-CodeQuality

gitsac committed:
Merge pull request #3 from gitsac/branch-A-CodeQuality

Improve code quality by refactoring.

gitsac committed:
Fix typo in Task class method

gitsac committed:
Add C-Tagging functionality

gitsac committed:
Improve GUI of chatbot

gitsac committed:
Center profile picture of user and chatbot

gitsac committed:
Fix background image for chat bot

gitsac committed:
Add unique personality to chatbot

gitsac committed:
Change response when tagging task

gitsac committed:
Add Ui.png

gitsac committed:
Change Ui.png to be more reflective

gitsac committed:
Update README.md

gitsac committed:
Update README.md

gitsac committed:
Fix checkstyle and test issues
